CorfuQueue: Persisted Queue over State Machine Replication

hisundar opened the pull request:
## Overview

Description:
A persisted queue over CorfuDB using state machine replication.
String append(Object) : Generates a non-conflicting UUID and possibly inserts into a CorfuTable
Object remove(id): Removes the object using the id
List<Map.Entry<String, Object>> entryList() returns a list of all the entries along with their ids, sorted by their realized stream addresses (thus logically a FIFO queue).

Why should this be merged:
An interface that now supports ordered insertion and retrieval without any fundamental alterations.

Related issue(s) (if applicable): #


## Checklist (Definition of Done):

- [x] There are no TODOs left in the code
- [x] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [ ] Change is covered by automated tests
- [x] Public API has Javadoc

medhavidhawan commented on runtime/src/main/java/org/corfudb/runtime/collections/iSMRDeque.java:
```diff
@@ -0,0 +1,139 @@
+ * Double Ended Queue immutable supported by CorfuDB using State Machine Replication.
+ * Created by Sundar Sridharan on 5/8/19.
+ */
+public interface iSMRDeque<E> extends Deque<E>, ISMRObject {
```
We do not want to implement Java.util interfaces. It becomes very onerous to maintain compatibility.

hisundar replied:
Done. Removed.

medhavidhawan commented on runtime/src/main/java/org/corfudb/runtime/collections/iSMRDeque.java:
```diff
@@ -0,0 +1,139 @@
+ * Double Ended Queue immutable supported by CorfuDB using State Machine Replication.
+ * Created by Sundar Sridharan on 5/8/19.
+ */
+public interface iSMRDeque<E> extends Deque<E>, ISMRObject {
```
Naming can follow ICorfuTable.

hisundar replied:
Done.

medhavidhawan commented on runtime/src/main/java/org/corfudb/runtime/collections/iSMRDeque.java:
```diff
@@ -0,0 +1,139 @@
+ * Double Ended Queue immutable supported by CorfuDB using State Machine Replication.
+ * Created by Sundar Sridharan on 5/8/19.
+ */
+public interface iSMRDeque<E> extends Deque<E>, ISMRObject {
```
Need more comments to it's intended use cases and also it's interaction with CorfuTable.

hisundar replied:
Done.

lixinchengdu commented on runtime/src/main/java/org/corfudb/runtime/collections/iSMRDeque.java:
```diff
@@ -0,0 +1,139 @@
+     */
+    @Accessor
+    Iterator<E> iterator();
+}
```
line 139 should be taken care of

hisundar replied:
done. thanks

vjeko commented on docs/corfu-queue-design.md:
```diff
@@ -0,0 +1,12 @@
+The class `CorfuQueue` implements a persisted queue over the abstraction of a CorfuTable. Currently CorfuTable uses a HashMap<> to represent the materialized view of a distributed map. A HashMap does not carry a notion of ordering natively. However when implemented over the abstraction of a distributed shared log, the elements added to the map do, in fact, have an ordering imposed by their append ...
+
+###  1. `String enqueue(E)`
+Since a map has a key and value, where key is a conflict parameter, enqueue simply generates a non-conflicting UUID as the key and inserts the Entry as a value into a CorfuTable. The generated UUID is returned for the caller.
```
How do we plan on generating non-conflicting UUIDs? It either that we will need something similar to Twitter Snowflake, or rely on the sequencer for the uniqueness.

hisundar replied:
That is a good question. For the prototype I was thinking of using Java UUID which as a collision probability of 1 in a billion
https://en.wikipedia.org/wiki/Universally_unique_identifier#Collisions
Since in our system we already have a sequencer which issues tokens and tokens also need to have the same property of uniqueness, I was hoping we can use tokens without breaking the abstraction.

vjeko replied:
Now that I think about it, even conflicting UUIDs should be fine, as along as you are able to detect the collision and re-generate the UUID.

Maithem replied:
Its actually not fine. Collision means data loss in this case. 

Consider this case: two clients successfully enqueue two items, which generates two UUIDs u1 and u2 for write1 and write2. Those IDs are then published, on the dequeue (i.e. the consumer), two different consumers cannot map u1 -> write1 and u2 -> write2 since u1 and u2 are equal, so they must be unique on generation.

vjeko commented on docs/corfu-queue-design.md:
```diff
@@ -0,0 +1,12 @@
+Since a map has a key and value, where key is a conflict parameter, enqueue simply generates a non-conflicting UUID as the key and inserts the Entry as a value into a CorfuTable. The generated UUID is returned for the caller.
+
+### 2. `List<Map.Entry<String, Object>> entryList()`
+Returns a list of all the entries along with their ids, sorted by their realized stream addresses (thus logically a FIFO queue). Assuming insertion order is preserved, CorfuTable can materialize its state as a [LinkedHashMap](https://docs.oracle.com/javase/8/docs/api/java/util/LinkedHashMap.html) instead of a HashMap and return the view as a list.
```
What does this mean WRT undo records. Do we need to keep them around, or do we just remove the head of the list?

hisundar replied:
If you mean the entryList() api then that is just a snapshot of a point in time. No undo records should be necessary.

Maithem replied:
undo records are generated for mutations during syncing. They need to be generated for mutators, but there is  no explicit control over them, they are cached in the address space.

hisundar commented:
Latest patch implements offline comments from @xnull and @medhavidhawan

xnull commented on runtime/src/main/java/org/corfudb/runtime/collections/ICorfuQueue.java:
```diff
@@ -0,0 +1,94 @@
+ * Created by Sundar Sridharan on 5/8/19.
+ * @param E entry to enqueue into the persisted queue
+ */
+public interface ICorfuQueue<E> extends ISMRObject {
```
please don't use `I` prefix

hisundar replied:
done. (No comments on the abundance of I in interface definitions in existing code :P)

hisundar replied:
Also I think I will get rid of the pure interface itself since there is only one class that is going to implement it. [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) principle.

xnull commented on runtime/src/main/java/org/corfudb/runtime/collections/ICorfuQueue.java:
```diff
@@ -0,0 +1,94 @@
+     * @return List of Entries sorted by their enqueue order
+     */
+    @Accessor
+    List<Map.Entry<EntryId, E>> entryList(int maxEntires);
```
Why we need this method? We can use `for cycle` and collect the same list by ourself. 
I also don't like that we return a particular data structure.  Why if I want to have sorted set instead of a list?
Also always try to avoid using Map.Entry, it's a raw structure doesn't tell anything about your domain entity, it's not convenient to use the entry. If you really need something like entry just create your own class that will describe your domain entry. For instance:
```
class CorfuRecord<E>{
  EntryId id;
  E value;
}
```

hisundar replied:
Done. will add the CorfuRecord<E>. But not sure what you mean about using "for cycle"? We need some method to return the elements (be it sorted set or list) right? So what will that method be?

hisundar commented on test/src/test/java/org/corfudb/runtime/collections/CorfuQueueTest.java:
```diff
@@ -0,0 +1,55 @@
+
+    @Test
+    @SuppressWarnings("unchecked")
+    public void basicQueueOrder() {
```
Sure. I am already working on those.

hisundar commented on test/src/test/java/org/corfudb/runtime/collections/CorfuQueueTest.java:
```diff
@@ -0,0 +1,55 @@
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class CorfuQueueTest extends AbstractViewTest {
```
Done.

annym commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java:
```diff
@@ -0,0 +1,207 @@
+     * <p>This function currently does not return a view like the java.util implementation,
+     * and changes to the entryList will *not* be reflected in the map. </p>
+     *
+     * @param maxEntires - Limit the number of entries returned from start of the queue
```
typo? maxEntries?

hisundar replied:
thanks. fixed.

annym commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java:
```diff
@@ -0,0 +1,207 @@
+    }
+
+    /**
+     * Erase all entries from the Queue.
```
Can we rephrase the 'erase'? Maybe remove.

hisundar replied:
Done.

WenbinZhu commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java:
```diff
@@ -0,0 +1,207 @@
+import org.corfudb.annotations.MutatorAccessor;
+import org.corfudb.runtime.view.OrderedGuidGenerator;
+
+import java.util.*;
```
Expand

hisundar replied:
done.

Maithem replied:
https://stackoverflow.com/questions/3587071/disable-intellij-starred-package-imports

WenbinZhu commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java:
```diff
@@ -0,0 +1,207 @@
+@Slf4j
+@CorfuObject
+public class CorfuQueue<E> implements ISMRObject {
+    /** The "main" linked map which contains the primary key-value mappings. */
```
Can you reformat all the comments in this patch? Single line comment should start with `//`, while multi-line comment should be 
```
/**
 * Start comment here.
 */
```

hisundar replied:
Done.

WenbinZhu commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java:
```diff
@@ -0,0 +1,207 @@
+        return id;
+    }
+
+    /** CorfuQueueRecord encapsulates each entry enqueued into CorfuQueue with its unique ID.
```
Inconsistent comment style.

hisundar replied:
Fixed. thanks.

WenbinZhu commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java:
```diff
@@ -0,0 +1,207 @@
+        return mainMap.isEmpty();
+    }
+
+    /** {@inheritDoc} */
```
```
/**
 * {@inheritDoc}
 */
```

hisundar replied:
done. thanks

WenbinZhu commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java:
```diff
@@ -0,0 +1,207 @@
+ */
+@Slf4j
+@CorfuObject
+public class CorfuQueue<E> implements ISMRObject {
```
No sure if `implements ISRObject` really needed here?

Maithem replied:
Since CorfuQueue is just a wrapper is it still needed?
CorfuQueue is itself not a SMR object, but wraps one.

hisundar replied:
done removed the ISMR

WenbinZhu commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java:
```diff
@@ -0,0 +1,207 @@
+     *                   -1 implies all entries
+     * @return List of Entries sorted by their enqueue order
+     */
+    public List<CorfuQueueRecord<E>> entryList(int maxEntries) {
```
In my understanding, this API is mostly used in recovery, meaning if the client crashed, lost all its in-memory Guid keys that are not consumed yet, this API helps recover it. Is my understanding correct? If so could you add it to the comment?

hisundar replied:
hi Wenbin, it is not used for recovery alone. This is the main "consumer" of the queue, which returns the entries as an ordered list. I will make this clearer in the Javadoc as you suggest..

WenbinZhu commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java:
```diff
@@ -0,0 +1,207 @@
+    }
+
+    /**
+     * Removes a specific element identified by the ID returned via entryList()'s CorfuQueueRecord.
```
The comment says `returned via entryList()'s CorfuQueueRecord`, but the implementation doesn't seem to use `CorfuQueueRecord`?

hisundar replied:
The caller must only supply the id part of the CorfuRecord() so we don't really need the CorfuRecord here right?

WenbinZhu replied:
I see, so this comment needs to be changed, right?

WenbinZhu commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java:
```diff
@@ -0,0 +1,207 @@
+     *
+     * {@inheritDoc}
+     *
+     * @throws NoSuchElementException if this queue did not contain this element
```
The underlying data structure is a map, will it throw `NoSuchElementException`?

WenbinZhu commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java:
```diff
@@ -0,0 +1,207 @@
+        undoRemove(queue, undoRecord, key);
+    }
+
+    enum UndoNullable {
```
Is this used?

hisundar replied:
No, I will get rid of it. thanks

WenbinZhu commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java:
```diff
@@ -0,0 +1,207 @@
+        mainMap.clear();
+    }
+
+    /** Helper function to get a Corfu Queue.
```
Comment style, please check others.

hisundar replied:
Copy paste from CorfuTable. But I will implement it as you suggest. thanks

WenbinZhu commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java:
```diff
@@ -0,0 +1,207 @@
+     */
+    static <E>
+    TypeToken<CorfuQueue<E>>
+        getQueueType() {
```
Is this used?

hisundar replied:
yes, it is used by the ObjectBuilder()

WenbinZhu commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java:
```diff
@@ -0,0 +1,207 @@
+        List<CorfuQueueRecord<E>> copy = new ArrayList<>(Math.min(mainMap.size(), maxEntries));
+        for (Map.Entry<UUID, E> entry : mainMap.entrySet()) {
+            copy.add(new CorfuQueueRecord<>(entry.getKey(),
+                    entry.getValue()));
```
I think you need to break the line here

hisundar replied:
done.

WenbinZhu commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java:
```diff
@@ -0,0 +1,207 @@
+     *
+     * @param e the element to add
+     * @throws NullPointerException if the specified element is null and this
+     *         queue does not permit null elements
```
Is there an option of permitting null elements or not?

hisundar replied:
I think null can be permitted. I will update the comment accordingly. thanks

WenbinZhu commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java:
```diff
@@ -0,0 +1,207 @@
+    /**
+     * Appends the specified element at the end of this unbounded queue.
+     * In a distributed system, the order of insertions cannot be guaranteed
+     * unless a transaction is used.
```
Could you explain why using a transactional `enque` can guarantee order? Because in my understanding, optimistic transaction checks key conflict on commit, while keys (Guid) are generated independently, and checks are only on commit, how is order being guaranteed?

hisundar replied:
Very good question Wenbin. There are 2 concerns here:
1. Identifying the elements (which the ordered guid generator is solving) and
2. Returning the elements in the same order in which the transactions committed
Currently the design only solves the first problem and provides a best effort for 2.
I am still working on ironing the out the requirements for guaranteeing the second part.

WenbinZhu commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java:
```diff
@@ -0,0 +1,207 @@
+ * Persisted Queue supported by CorfuDB using distributed State Machine Replication.
+ * Entries enqueued are backed by a LinkedHashMap where each entry is mapped to a unique and
+ * globally ordered generated id that is returned upon successful <b>enqueue()</b>.
+ * <b>entryList()</b> returns the enqueued entries along with their ids sorted by
```
Seems like the implementation of entryList() does not do any sorting?

hisundar replied:
It does not. The ordering is something I am still working on getting as per expectations.

hisundar replied:
With transaction snapshot + index in the queue, I believe we should have global ordering. Could you please let me know if this looks ok? thanks

WenbinZhu commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java:
```diff
@@ -0,0 +1,207 @@
+                break;
+            }
+        }
+        return copy;
```
I think the result should be sorted, right?

hisundar replied:
I was trying to see if we can avoid the sorting since the SMR will anyway apply the changes in the same order in which they appear in the stream right? So as mentioned in the design doc above, having a LinkedHashMap should suffice?

hisundar commented:
> Please use the snapshot version and internal insertion order to implement the comparison functionality.

Sure, working on that.

Maithem commented on runtime/src/main/java/org/corfudb/runtime/view/ObjectBuilder.java:
```diff
@@ -46,6 +46,9 @@
     ISerializer serializer = Serializers.JSON;
 
+    @Setter
+    boolean needOrderedUUIDs;
```
This seems to be specific to the object type itself. Needs to be pushed to the constructor.

hisundar replied:
Done. I will push this out to the constructor as you suggest.

Maithem commented on runtime/src/main/java/org/corfudb/runtime/view/ObjectBuilder.java:
```diff
@@ -92,6 +95,11 @@ public T open() {
         }
 
+        // Some Corfu Collections like CorfuQueue need globally ordered UUIDs
+        if (needOrderedUUIDs) {
```
This can be instrumented from a higher layer. The object view doesn't need to know whether the uuid are ordered or not, it just needs a uuid.

hisundar replied:
Done.

Maithem commented on docs/corfu-queue-design.md:
```diff
@@ -0,0 +1,15 @@
+The class `CorfuQueue` implements a persisted queue over the abstraction of a CorfuTable. Currently CorfuTable uses a HashMap<> to represent the materialized view of a distributed map. A HashMap does not carry a notion of ordering natively. However when implemented over the abstraction of a distributed shared log, the elements added to the map do, in fact, have an ordering imposed by their append ...
```
This doesn't really use the `CorfuTable` and the FIFO order is dictated by the global commit order.

hisundar replied:
Good point. Updating this doc now.

Maithem commented on docs/corfu-queue-design.md:
```diff
@@ -0,0 +1,15 @@
+Assuming checkpointing and garbage collection also work in the same insertion order, this proposal is to implement the abstraction of a logical immutable queue using nothing more than a [LinkedHashMap](https://docs.oracle.com/javase/8/docs/api/java/util/LinkedHashMap.html).
+
+### A Note about the UUIDs
+The UUIDs are the keys of the underlying [LinkedHashMap](https://docs.oracle.com/javase/8/docs/api/java/util/LinkedHashMap.html). These UUIDs will need to have be globally unique and also comparable to facilitate easy processing of the queue elements. The unique UUIDs can be simply generated using a epoch aware counter inside the Sequencer.
```
This seems a little out dated.

hisundar replied:
Good catch. Let me update this too. thanks

Maithem commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java:
```diff
@@ -0,0 +1,300 @@
+     *          The ID returned here is only really useful for remove() operations.
+     */
+    @TransactionalMethod
+    public UUID enqueue(E e) {
```
Removing the SMR counter will simplify this. Nested transactions per enqueue can be expensive and really its not needed.

Maithem replied:
Hmm, it seems like annotation processor is actually broken and TransactionalMethod is needed.

Maithem commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java:
```diff
@@ -0,0 +1,300 @@
+     *
+     * @param <E>
+     */
+    public static class CorfuQueueRecord<E> {
```
If you return an ordered map, you can get rid of this class. I don't see a strong reason to transform the map to a List when the map is already ordered.

hisundar replied:
There are a few reasons:
1. The Map internally is a long=>entry, to return something consumable externally it will have to be remapped to an UUID=>entry, so the internal map to external consumption still needs an iterator.
2. I was addressing comments from @xnull who asked me to add this class to avoid overloading the Map.

Maithem commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java:
```diff
@@ -0,0 +1,300 @@
+     */
+    @Accessor
+    protected boolean containsKeyInternal(@ConflictParameter Long key) {
+        return mainMap.containsKey(key);
```
this should be a private helper method (remove the @Accessor)

hisundar replied:
done.

Maithem commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java:
```diff
@@ -0,0 +1,300 @@
+    /**
+     * {@inheritDoc}
+     */
+    @TransactionalMethod(readOnly = true)
```
This seems to be just an `@Accessor` , also don't forget to include the `@ConflictParameter` annotation

hisundar replied:
Again, how do we specify @ConflictParameter on a part of the UUID? The UUID itself cannot be the conflict parameter right?

Maithem commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java:
```diff
@@ -0,0 +1,300 @@
+     * {@inheritDoc}
+     */
+    @Accessor
+    protected E getInternal(@ConflictParameter Long key) {
```
The annotations should be on the public API, not the internal helper methods.

Maithem commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java:
```diff
@@ -0,0 +1,300 @@
+     * {@inheritDoc}
+     */
+    @Accessor
+    public boolean containsKey(UUID key) {
```
Need the `@ConflictParameter`

hisundar replied:
How do I specify a @ConflictParameter on just the least significant bytes of an UUID?
That's why its currently separated into a public method calling an internal method with an @Accessor.

xnull replied:
BTW, you have to double check that `@Accessor` works correctly in this case.  There is a limitation (at least in hibernate/spring I have seen it), when you try to execute an annotated method you expect the logic related to the annotation will be triggered, which not happens, unfortunately (there is a limitation in Java annotation processing)

hisundar replied:
Ok @xnull let me re-write the CorfuQueue completely as per your suggestion :)

Maithem commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java:
```diff
@@ -0,0 +1,300 @@
+     * @return The entry that was successfully removed or null if there was no mapping.
+     */
+    @TransactionalMethod
+    public E removeEntry(UUID id) {
```
Need to annotate this with a `MutatorAccessor` and mark the parameter as a `@ConflictParameter`

Maithem replied:
Is this method still needed?

Maithem commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java:
```diff
@@ -0,0 +1,300 @@
+    /**
+     * {@inheritDoc}
+     */
+    @MutatorAccessor(name = "put", undoFunction = "undoPut", undoRecordFunction = "undoPutRecord")
```
Why not put these on the public API and not the internal data structure?

Maithem commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java:
```diff
@@ -0,0 +1,300 @@
+    @MutatorAccessor(name = "remove", undoFunction = "undoRemove",
+                                undoRecordFunction = "undoRemoveRecord")
+    @SuppressWarnings("unchecked")
+    protected E remove(@ConflictParameter Long key) {
```
I see, since the conflict parameter is generated internally its pushed downwards.

Maithem commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java:
```diff
@@ -0,0 +1,300 @@
+        }
+
+        long snapshotVersion = 0;
+        if (TransactionalContext.getCurrentContext() != null) {
```
This is incorrect. Since snapshot version 0 and a non transactional access would use the same snapshot, but the object can be different (i.e. collisions). I would just rely on the VLO version at the time of the access.

Maithem commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java:
```diff
@@ -0,0 +1,300 @@
+        if (TransactionalContext.getCurrentContext() != null) {
+            snapshotVersion = TransactionalContext.getCurrentContext().getSnapshotTimestamp().getSequence();
+        }
+        List<CorfuQueueRecord<E>> copy = new ArrayList<>(Math.min(mainMap.size(), maxEntries));
```
the returned collection reflects commit order, then why does the UUID itself have to have time ordering?
It seems like the consumer should maintain that.

hisundar replied:
Having the UUID allows easier comparison across snapshots. This was an explicit ask.

Maithem commented on docs/corfu-queue-design.md:
```diff
@@ -0,0 +1,12 @@
+The class `CorfuQueue` implements a persisted queue over the abstraction of a CorfuTable. Currently CorfuTable uses a HashMap<> to represent the materialized view of a distributed map. A HashMap does not carry a notion of ordering natively. However when implemented over the abstraction of a distributed shared log, the elements added to the map do, in fact, have an ordering imposed by their append ...
```
HashMap -> LinkedHashMap

hisundar replied:
done.

Maithem commented on docs/corfu-queue-design.md:
```diff
@@ -0,0 +1,12 @@
+The class `CorfuQueue` implements a persisted queue over the abstraction of a CorfuTable. Currently CorfuTable uses a HashMap<> to represent the materialized view of a distributed map. A HashMap does not carry a notion of ordering natively. However when implemented over the abstraction of a distributed shared log, the elements added to the map do, in fact, have an ordering imposed by their append ...
+
+###  1. `UUID enqueue(E)`
+Since a map has a key and value, where key is a conflict parameter, enqueue generates a non-conflicting Long as the key and inserts the Entry as a value into a CorfuTable. The generated Long is packed into the LSB of a UUID and returned to the caller.
```
Should we also note the fact that the UUID returned by enqueue is different from the one returned by entryList? Since two different UUIDs can reference the same entry it seems like a weird semantic that we should be clear about (i.e. these are "special uuids")

hisundar replied:
The api does note it clearly, but let me add it to the design doc as well.

Maithem commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java:
```diff
@@ -0,0 +1,224 @@
+     *          The ID returned here is only really useful for remove() operations.
+     */
+    public UUID enqueue(E e) {
+        final Long id = runtime.getGuidGenerator().nextLong();
```
The runtime is already overloaded with so misplaced functionality. I don't think the OrderedGuidGenerator should be part of it. We can pass a OrderedGuidGenerator through the constructor.

hisundar replied:
done.

Maithem commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java:
```diff
@@ -0,0 +1,224 @@
+     *                   -1 implies all entries
+     * @return List of Entries sorted by their enqueue order
+     */
+    public List<CorfuQueueRecord<E>> entryList(int maxEntries) {
```
I think its cleaner to overload instead of pass a -1. 

`public List<CorfuQueueRecord<E>> entryList() ` returns all elements

`public List<CorfuQueueRecord<E>> entryList(int x) ` returns first x elements

hisundar replied:
done.

Maithem commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java:
```diff
@@ -0,0 +1,224 @@
+
+        long snapshotVersion;
+        boolean startedNewTransaction = false;
+        if (TransactionalContext.isInTransaction()) {
```
I suggest you factor out the getInternalEntrySet logic into

```java

    Pair<Long, Set<Map.Entry<Long, E>>> getEntries() {
        final Set<Map.Entry<Long, E>> res;
        final long version;

        if (TransactionalContext.isInTransaction()) {
            res = corfuTable.entrySet();
            version = TransactionalContext.getCurrentContext().getSnapshotTimestamp().getSequence();
        } else {
            runtime.getObjectsView().TXBuild()
                    .build()
                    .begin();
            try {
                res = corfuTable.entrySet();
                version = TransactionalContext.getCurrentContext().getSnapshotTimestamp().getSequence();
            } finally {
                runtime.getObjectsView().TXEnd();
            }
        }
        return new Pair<>(version, res);
    }
```

hisundar replied:
Sure, but are we guaranteed that the ordering will be preserved from the List->Set->List conversions?

hisundar replied:
Also would like consensus with @xnull 's comments in the same PR who had explicitly discouraged the use of Map.Entry<>.

xnull replied:
yes, I am)

Maithem commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuTable.java:
```diff
@@ -267,26 +267,39 @@ public ProjectionFunction getProjectionFunction() {
+    /**
+     * Generate a table with a given implementation for the mainMap
+     */
+    public CorfuTable(IndexRegistry<K, V> indices, Map<K, V> mapImpl) {
```
This might break existing code, just overload this.

Maithem commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuTable.java:
```diff
@@ -267,26 +267,39 @@ public ProjectionFunction getProjectionFunction() {
+    /**
+     * Generate a table with a given implementation for the mainMap
+     */
+    public CorfuTable(IndexRegistry<K, V> indices, Map<K, V> mapImpl) {
```
It might be cleaner to pass a type and not an instance. 

https://stackoverflow.com/questions/3574065/instantiate-a-class-object-with-constructor-that-accepts-a-string-parameter

hisundar replied:
I thought about this, but was wondering about passing custom implementation of the map. For example if we pass only a LinkedHashMap or HashMap as a type and have CorfuTable instantiate it, that will limit us from passing in some other Map which can do other things. Any thoughts?

xnull commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java:
```diff
@@ -0,0 +1,226 @@
+
+    public CorfuQueue(CorfuRuntime runtime, String streamName) {
+        this.runtime = runtime;
+        corfuTable = runtime.getObjectsView().build()
```
can you please pass CorfuTable directly instead of CorfuRuntime

hisundar replied:
CorfuTable is internal to the implementation of this data structure. Runtime is needed not just for building the CorfuTable but also for transactions and snapshot versions. Please suggest an alternative if not for runtime?

xnull commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java:
```diff
@@ -0,0 +1,226 @@
+                .setStreamName(streamName)
+                .setArguments(CorfuTable.IndexRegistry.empty(), new LinkedHashMap<Long, E>())
+                .open();
+        guidGenerator = new CorfuGuidGenerator(runtime);
```
It would be great to pass StreamView or something like that, but not CorfuRuntime

hisundar replied:
StreamView does not allow access to transactions. We need access to transactions, snapshot versions as well as a way to initialize CorfuTable. runtime provides all of this, please suggest an alternative if not runtime.

xnull commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java:
```diff
@@ -0,0 +1,226 @@
+     *          The ID returned here is only really useful for remove() operations.
+     */
+    public UUID enqueue(E e) {
+        final Long id = guidGenerator.nextLong();
```
can we create a Dto class which would contain a UUID and a Long or something like that?
Using longs and uuids confusing

hisundar replied:
That's what the first version had (CorfuQueueId), but I was asked to use UUID since consumers are using that. Since both you and @Maithem have asked for a separate type, I don't mind putting CorfuQueueId back, but I don't know the repercussions.

xnull replied:
To me - it had better use that class, just mark old API as deprecated. This mess with longs and UUIDs confusing

hisundar replied:
done. using a custom class.

xnull commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java:
```diff
@@ -0,0 +1,226 @@
+         * because if this method is wrapped in a transaction, the order is established only later.
+         */
+        @Getter
+        final UUID ID;
```
private

hisundar replied:
done.

xnull commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java:
```diff
@@ -0,0 +1,226 @@
+        final UUID ID;
+
+        @Getter
+        final E entry;
```
private

hisundar replied:
done.

xnull commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java:
```diff
@@ -0,0 +1,226 @@
+        final E entry;
+
+        public String toString() {
+            return String.format(ID.toString()+"=>"+entry.toString());
```
code format

hisundar replied:
done.

xnull commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java:
```diff
@@ -0,0 +1,226 @@
+        }
+    }
+
+    final private static int MAX_BITS_FOR_INDEX = 24;
```
comment please

hisundar replied:
done.

xnull commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java:
```diff
@@ -0,0 +1,226 @@
+     */
+    public List<CorfuQueueRecord<E>> entryList(int maxEntries) {
+        if (maxEntries < 0) {
+            maxEntries = Integer.MAX_VALUE;
```
please add logs - suspicious value. Why it less than zero?

hisundar replied:
done. better to throw illegal exception I think

xnull replied:
My advice is  instead of throwing an Exception use https://github.com/CorfuDB/CorfuDB/blob/master/common/src/main/java/org/corfudb/common/result/Result.java

xnull commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java:
```diff
@@ -0,0 +1,226 @@
+            runtime.getObjectsView().TXBuild().type(TransactionType.WRITE_AFTER_WRITE)
+                    .build()
+                    .begin();
+            snapshotVersion = TransactionalContext.getCurrentContext().getSnapshotTimestamp().getSequence();
```
can we pass `TransactionalContext.getCurrentContext()` as a dependency?

hisundar replied:
what if the caller does not want to use transactions? Also transactions is an internal implementation detail, we cannot ask caller of this function to provide something internal right?

xnull commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java:
```diff
@@ -0,0 +1,226 @@
+        if (startedNewTransaction) {
+            runtime.getObjectsView().TXEnd();
+        }
+        return copy;
```
can we return an immutable list?

hisundar replied:
we do internally. We return an ImmutableListSetWrapper, please check out https://github.com/CorfuDB/CorfuDB/blob/master/runtime/src/main/java/org/corfudb/runtime/collections/CorfuTable.java#L605

xnull commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuTable.java:
```diff
@@ -267,26 +267,39 @@ public ProjectionFunction getProjectionFunction() {
     private final Map<String, Map<Comparable, Map<K, V>>> secondaryIndexes = new HashMap<>();
 
     @Getter
     boolean indexGenerationFailed = false;
```
private and final if possible

hisundar replied:
CorfuTable has several things that are non-conformant. I explicitly refrained from addressing those in this PR. We should have a separate PR addressing those since there are so many.

xnull commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuTable.java:
```diff
@@ -267,26 +267,39 @@ public ProjectionFunction getProjectionFunction() {
+    /**
+     * Generate a table with a given implementation for the mainMap
+     */
+    public CorfuTable(IndexRegistry<K, V> indices, Map<K, V> mapImpl) {
```
parameter name `mapImpl` completle unclear

hisundar replied:
I would refrain from changing CorfuTable in this PR and have that as a separate change. There are a lot more comments I am sure on CorfuTable than what you see here :)

xnull commented on test/src/test/java/org/corfudb/runtime/concurrent/CorfuQueueTxTest.java:
```diff
@@ -0,0 +1,91 @@
+        ArrayList<Record> validator = new ArrayList<>(numThreads * numIterations);
+        ReentrantLock lock = new ReentrantLock();
+
+        scheduleConcurrently(numThreads, t -> {
```
please reformat code

hisundar replied:
done.

vjeko commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java:
```diff
@@ -0,0 +1,226 @@
+     *     +----------------------------------------------------------------------------+
+     *     | Stream snapshot |Index in snapshot|     ID of the entry in the map        |
+     *     +----------------------------------------------------------------------------+
+     *     <----5 bytes------><----3 bytes-----><------- 8 bytes ----------------------->
```
What is the use case for having the UUID comparable? The fact we have one type and two semantic definitions makes it very confusing.

hisundar replied:
Agreed. I have changed that in the latest patchiest. Could you please review. We do have a need for returning a global comparable ordering via an Id. Having it as an UUID makes it confusing, so I have encapsulated it as a separate DTO with constraints on comparability.

vjeko commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java:
```diff
@@ -0,0 +1,226 @@
+            copy.add(record);
+        }
+        // Given that we are using a WRITE_AFTER_WRITE on a read-only txn, we expect no aborts.
+        if (startedNewTransaction) {
```
You do not geed state for this functionality. Why not:

```
private entryListImpl() {
  ...
}

public entryList() {

  if (!inTransaction) {
    return entryList();
  }

  startTx();
  entryList();
  endTx();
}
```

hisundar replied:
The transaction is not for the snapshot of the entryList() that is already protected by the VLO.
The transaction is a means of obtaining and binding the snapshot version of the list with its index. I have updated the comments on the Java doc in the latest upload explaining the same.

vjeko commented on test/src/test/java/org/corfudb/runtime/collections/CorfuQueueTest.java:
```diff
@@ -0,0 +1,52 @@
+import org.corfudb.runtime.view.AbstractViewTest;
+import org.junit.Test;
+
+import java.util.*;
```
Do not use star import.

hisundar replied:
done.

vjeko commented on test/src/test/java/org/corfudb/runtime/concurrent/CorfuQueueTxTest.java:
```diff
@@ -0,0 +1,91 @@
+import org.corfudb.runtime.object.transactions.AbstractTransactionsTest;
+import org.junit.Test;
+
+import java.util.*;
```
Same here.

hisundar replied:
done. thanks

vjeko commented on test/src/test/java/org/corfudb/runtime/concurrent/CorfuQueueTxTest.java:
```diff
@@ -0,0 +1,91 @@
+                    final long streamOffset = TXEnd();
+                    validator.add(new Record(id, queueData));
+                    log.debug("ENQ:"+id+"=>"+queueData+" at "+streamOffset);
+                    lock.unlock();
```
You can move this to finally.

hisundar replied:
done. thanks

hisundar commented on docs/corfu-queue-design.md:
```diff
@@ -0,0 +1,12 @@
+Since a map has a key and value, where key is a conflict parameter, enqueue generates a non-conflicting Long as the key and inserts the Entry as a value into a CorfuTable. The generated Long is packed into the LSB of a UUID and returned to the caller.
+
+### 2. `List<CorfuQueueRecord<Object>> entryList()`
+Returns a list of all the entries along with their ids. These ids are packed into UUIDs with the MSB bits composed out of their (stream snapshot address, index in queue) tuples and their LSB bits with the enqueued Long. Thus these Ids have a global comparable ordering maintained by their underlying CorfuTable's [LinkedHashMap](https://docs.oracle.com/javase/8/docs/api/java/util/LinkedHashMap.html ...
```
Agreed. @xnull and @Maithem had similar comments as well. I have changed it back to a custom type called CorfuRecordId which makes this intricacy internal and validates incorrect comparisons.

hisundar commented:
Hi @xnull  @Maithem @vjeko I have addressed your comments and also added extra test cases for validating some of your concerns (like ordering and comparability of the IDs) could you please take a look whenever possible? thanks!

lixinchengdu commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java:
```diff
@@ -0,0 +1,298 @@
+        }
+
+        @Override
+        public int compareTo(CorfuQueueRecord<E> o) {
```
Is it better to have the parameter of compareTo being CorfuQueueRecord<? extends E>?

hisundar replied:
this is a great suggestion! Thank you!

lixinchengdu commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java:
```diff
@@ -0,0 +1,298 @@
+     * @throws IllegalArgumentException if maxEntries is negative.
+     * @return List of Entries sorted by their enqueue order
+     */
+    public List<CorfuQueueRecord<E>> entryList(int maxEntries) {
```
Is it possible to have List<CorfuQueueRecord<? super E>>?

hisundar replied:
great suggestion again. thank you

lixinchengdu commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java:
```diff
@@ -0,0 +1,298 @@
+     * @return List of Entries sorted by their enqueue order
+     */
+    public List<CorfuQueueRecord<E>> entryList(int maxEntries) {
+        if (maxEntries <= 0) {
```
Why 0 is not correct? BTW, the error message only says that maxEntries should not be negative.

hisundar replied:
Good catch. I will fix this. thanks

hisundar commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java:
```diff
@@ -0,0 +1,298 @@
+     * with their snapshot+index id (also a long) which represents a global comparable ordering.
+     * This class encapsulates these two longs into one Id and add rules on comparability.
+     */
+    public class CorfuRecordId implements Comparable<CorfuRecordId> {
```
Yes, that is a good suggestion. Let me make it so. thanks

hisundar commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java:
```diff
@@ -0,0 +1,298 @@
+
+    /**
+     * Appends the specified element at the end of this unbounded queue.
+     * In a distributed system, the order of insertions cannot be guaranteed
```
Yes. Let me make that specific.

hisundar commented:
thanks @lixinchengdu I have addressed your comments in the latest patch.

xnull commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java:
```diff
@@ -0,0 +1,304 @@
+         * because if this method is wrapped in a transaction, the order is established only later.
+         */
+        @Getter
+        final private CorfuRecordId recordId;
```
private final)

hisundar replied:
done.

xnull commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java:
```diff
@@ -0,0 +1,304 @@
+     *
+     * @param <E>
+     */
+    public static class CorfuQueueRecord<E> implements Comparable<CorfuQueueRecord<? extends E>> {
```
`Comparable<CorfuQueueRecord<E>>` ?

hisundar replied:
Comparable<CorfuQueueRecord<? extends E>> was in response to @lixinchengdu 's comments above. Please let me know if you agree? thanks

lixinchengdu replied:
I am not sure whether there is any pitfall by doing this way. I am checking whether other having the same issue .

xnull commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuTable.java:
```diff
@@ -267,26 +267,39 @@ public ProjectionFunction getProjectionFunction() {
         });
-        log.info("CorfuTable: creating CorfuTable with the following indexes: {}", secondaryIndexes.keySet().toString());
+        log.info("CorfuTable: creating CorfuTable with the following indexes: {}",
+                secondaryIndexes.keySet().toString());
```
toString() not needed here

hisundar replied:
thanks. fixed

hisundar replied:
fyi, I did not write that code :)

hisundar commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java:
```diff
@@ -0,0 +1,304 @@
+     */
+    public List<CorfuQueueRecord<E>> entryList(int maxEntries) {
+        if (maxEntries < 0) {
+            throw new IllegalArgumentException("entryList given negative maxEntries");
```
I was following comments from @lixinchengdu but I can change it as you suggest. thanks

hisundar commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java:
```diff
@@ -0,0 +1,304 @@
+            if (++index >= maxEntries) {
+                break;
+            }
+            long ordering = (snapshotVersion<<MAX_BITS_FOR_INDEX) | index;
```
sure, it would be good to have an additional check, I will add it. thanks
(Note that 24 bits means 16.7 million entries in a single table. Pretty sure we will hit other limits in the system before we can return 16.7million records all at once in a single list)

hisundar commented on runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java:
```diff
@@ -0,0 +1,304 @@
+        // Bind the iteration order to a snapshot of the Queue using a transaction.
+        long snapshotVersion;
+        boolean startedNewTransaction = false;
+        if (TransactionalContext.isInTransaction()) {
```
Yes, thanks for the observation @WenbinZhu , please note that this is solved with the CorfuRecordId class's equals() function which only matches the LSB (the Id part), so in your example id1 from t1 and id2 from t2 will match to be equal. Also this is something we discussed quite a bit. The original plan was to return the exact stream addresses, but that would break layer abstraction, so it was decided to return ids in this manner.

WenbinZhu replied:
Yes, if we use euqals() I think it's fine, but if we care about  ordering and use compareTo(), <99, 1, id2> will be smaller than <100, 0, id1>. Does that break your semantics?

hisundar commented:
hi @WenbinZhu I have addressed your comments in the latest patch. Could you please let me know if it looks ok? thanks

hisundar commented:
> LGTM, maybe just put a comment as we discussed just now about transactions.

Thanks @WenbinZhu , I will add the comment and your exact example so folks are clear about the ordering of the entries returned.
